storage/innobase/xtrabackup/src/xbcloud/s3.cc

```diff
@@ -541,13 +541,19 @@ void S3_client::upload_callback(
     S3_client *client, std::string bucket, std::string name, Http_request *req,
     Http_response *resp, const Http_client *http_client, Event_handler *h,
     S3_client::async_upload_callback_t callback, CURLcode rc,
-    const Http_connection *conn, int count) {
+    const Http_connection *conn, ulong count) {
   bool retry_error = false;
-  if (retriable_curl_error(rc) ||
+  if (http_client->retriable_curl_error(rc) ||
       retriable_http_error(conn->response().http_code())) {
```

Comment:
is there a use case of giving user choice of http_error apart from curl_error?

Reply:
I have not faced any issue with it. If intermittent network I would expect that errors to be raised on curl side. If you wish, we can extent the options and create a new parameter for http codes.

Reply:
like for gc, we have http error 502 should be retry https://cloud.google.com/storage/docs/json_api/v1/status-codes. so I think it is good to give user choice similar to curl_error.

Reply:
ack, lets just agree on the item above (related to the initialization of http client) and then I will implement the http codes as well.

Reply:
> is there a use case of giving user choice of http_error apart from curl_error?

If it helps, in our use case the error we get on S3 downloads of very large backups is:
```
210310 04:41:29 xbcloud: Failed to upload object REDACTED-S3-BUCKET/REDACTED.ibd.qp.xbcrypt.00000000000000195146. Error message: We encountered an internal error. Please try again.
```
It appears this is a generic error AWS S3 returns for unknown reasons. Usually we don't see this error on smaller backups < 1TB

Reply:
currently we retry for HTTP erros 503, 500, 504, 408

Reply:
I see, so the above internal error message is definitely other than 503, 500, 504, 408
@timvaillancourt Can you please find out the error code from AWS documentation or from your logs?

Reply:
@satya-bodapati `xbcloud` doesn't report the status code so I cannot be sure, but Stackoverflow has examples of it being an HTTP 500 error that AWS returns: https://stackoverflow.com/questions/60437220/aws-s3-bucket-internal-server-error-500-on-post-file-upload-from-browser

We see this fail uploads to S3 on `percona-xtrabackup-24` version `2.4.22` due to this version having no backoff retry logic

Reply:
@rahulmalik87 Done